File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        def solve(root):
            if root.left==None and root.right==None:
                return [1,0]
            else:
                a=b=[0,0]
                if root.left:
                    a=solve(root.left)
                if root.right:
                    b=solve(root.right)
                heightmax=max(a[0],b[0])+1
                ansmax=max(a[0]+b[0],a[1],b[1])
                return [heightmax,ansmax]
        if root==None:
            return 0
        return solve(root)[1]
                    
        
        

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ans=0;
    int diameterOfBinaryTree(TreeNode* root) {
        solve(root);
        return ans;
    }
    int solve(TreeNode* root)
    {
        if(root==nullptr)return 0;
        else
        {
            int a,b;
            a=solve(root->left);
            b=solve(root->right);
            ans=max(ans,a+b);
            // cout<<root->val<<" "<<a<<" "<<b<<endl;
            return max(a,b)+1;
        }
    }
};